<a href="https://colab.research.google.com/github/chrisdionisius/MachineLearning2022/blob/main/Jobsheet06_NaiveBayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Tugas 1

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`
2. Fitur `CountVectorizer` dengan mengaktifkan **stop_words**
3. Evaluasi hasilnya

In [56]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [57]:
import numpy as np
import pandas as pd

In [74]:
# baca data
df = pd.read_csv('/content/drive/MyDrive/dataset/spam.csv', encoding='latin-1')

# Drop 3 kolom terakhir dengan fungsi iloc
df = df.drop(df.iloc[:,2:], axis=1)

# Cek data
df.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [75]:
# Data untuk rename kolom v1 dan v2
new_cols = {
    'v1': 'Labels',
    'v2': 'SMS'
}

# Rename nama kolom v1 dan v2
df = df.rename(columns=new_cols)

# cek data
df.head()

Labels                                                SMS
0    ham  Go until jurong point, crazy.. Available only ...
1    ham                      Ok lar... Joking wif u oni...
2   spam  Free entry in 2 a wkly comp to win FA Cup fina...
3    ham  U dun say so early hor... U c already then say...
4    ham  Nah I don't think he goes to usf, he lives aro...

In [76]:
# Cek Jumlah Data Per Kelas
print(df['Labels'].value_counts())
print('\n')

# Cek Kelengkapan Data
print(df.info())
print('\n')

# Cek Statistik Deskriptif
print(df.describe())

ham     4825
spam     747
Name: Labels, dtype: int64


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Labels  5572 non-null   object
 1   SMS     5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB
None


       Labels                     SMS
count    5572                    5572
unique      2                    5169
top       ham  Sorry, I'll call later
freq     4825                      30


In [77]:
# Data untuk label
new_labels = {
    'spam': 1,
    'ham': 0
}

# Encode label
df['Labels'] = df['Labels'].map(new_labels)

# Cek data
df.head()

Labels                                                SMS
0       0  Go until jurong point, crazy.. Available only ...
1       0                      Ok lar... Joking wif u oni...
2       1  Free entry in 2 a wkly comp to win FA Cup fina...
3       0  U dun say so early hor... U c already then say...
4       0  Nah I don't think he goes to usf, he lives aro...

In [78]:
X = df['SMS'].values
y = df['Labels'].values

In [94]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer

# Split data training dan testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)

# Inisiasi CountVectorizer
bow = CountVectorizer()


# Fitting dan transform X_train dengan CountVectorizer
X_train = bow.fit_transform(X_train)

# Transform X_test
X_test = bow.transform(X_test)

In [95]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

# Inisiasi MultinomialNB
mnb = MultinomialNB()

# Fit model
mnb.fit(X_train, y_train)

# Prediksi dengan data training
y_pred_train = mnb.predict(X_train)

# Evaluasi akurasi data training
acc_train = accuracy_score(y_train, y_pred_train)

# Prediksi dengan data training
y_pred_test = mnb.predict(X_test)

# Evaluasi akurasi data training
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print('Tanpa stop_word :')
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Tanpa stop_word :
Hasil akurasi data train: 0.9946152120260264
Hasil akurasi data test: 0.9775784753363229


In [89]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
bow = CountVectorizer(stop_words=["all","in","the","is","and"])
# bow = CountVectorizer(min_df=2)
# bow = CountVectorizer(max_df=0.50)
X_train = bow.fit_transform(X_train)
X_test = bow.transform(X_test)
mnb = MultinomialNB()
mnb.fit(X_train, y_train)
y_pred_train = mnb.predict(X_train)
acc_train = accuracy_score(y_train, y_pred_train)
y_pred_test = mnb.predict(X_test)
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print('Menggunakan stop_word :')
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Menggunakan stop_word :
Hasil akurasi data train: 0.9939421135292797
Hasil akurasi data test: 0.979372197309417


In [91]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
# bow = CountVectorizer(stop_words=["all","in","the","is","and"])
bow = CountVectorizer(min_df=2)
# bow = CountVectorizer(max_df=0.50)
X_train = bow.fit_transform(X_train)
X_test = bow.transform(X_test)
mnb = MultinomialNB()
mnb.fit(X_train, y_train)
y_pred_train = mnb.predict(X_train)
acc_train = accuracy_score(y_train, y_pred_train)
y_pred_test = mnb.predict(X_test)
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print('Menggunakan min_df :')
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Menggunakan min_df :
Hasil akurasi data train: 0.9903522548799641
Hasil akurasi data test: 0.97847533632287


In [96]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
# bow = CountVectorizer(stop_words=["all","in","the","is","and"])
# bow = CountVectorizer(min_df=2)
bow = CountVectorizer(max_df=0.50)
X_train = bow.fit_transform(X_train)
X_test = bow.transform(X_test)
mnb = MultinomialNB()
mnb.fit(X_train, y_train)
y_pred_train = mnb.predict(X_train)
acc_train = accuracy_score(y_train, y_pred_train)
y_pred_test = mnb.predict(X_test)
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print('Menggunakan max_df :')
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Menggunakan max_df :
Hasil akurasi data train: 0.9946152120260264
Hasil akurasi data test: 0.9775784753363229


## Evaluasi
dari 3 cara implementasi stop_word, 
* didapatkan penggunaan max_df memiliki akurasi data train yang paling tinggi yakni : 0.9946152120260264
* sementara untuk implementasi stop_word yang didefinisikan secara manual memiliki akurasi data test paling tinggi yakni : 0.979372197309417

### Catatan
Akurasi yang dihasilkan juga dipengaruhi kata-kata yang didefinisikan, bahasa, dan parameter pada function min_df/max_df (sumber : [kavita ganesan](https://kavita-ganesan.com/how-to-use-countvectorizer/#.YzQXfthBxhE))

## Tugas 2

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`
2. Fitur `TF-IDF` dengan mengaktifkan **stop_words**
3. Evaluasi hasilnya dan bandingkan dengan hasil tugas 1.
4. Berikan kesimpulan fitur mana yang terbaik pada kasus data `spam.csv`

In [87]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
bow = TfidfVectorizer(stop_words=["all","in","the","is","and"])
X_train = bow.fit_transform(X_train)
X_test = bow.transform(X_test)

In [88]:
mnb = MultinomialNB()
mnb.fit(X_train, y_train)
y_pred_train = mnb.predict(X_train)
acc_train = accuracy_score(y_train, y_pred_train)
y_pred_test = mnb.predict(X_test)
acc_test = accuracy_score(y_test, y_pred_test)

# Print hasil evaluasi
print('Menggunakan stop_word :')
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Menggunakan stop_word :
Hasil akurasi data train: 0.9771146511106125
Hasil akurasi data test: 0.947085201793722


## Kesimpulan

Pada tugas kali ini dilakukan klasfikasi Multinomial dengan 2 algoritma berbeda yakni CountVectorizer dan TfidfVectorizer, dimana ditambahkan juga *stop word* 
jika dilihat dari tingkat akurasi dapat dikatakan hasil indexing dari CountVectorizer lebih tinggi dibanding menggunakan TfidfVectorizer namun mungkin hal ini disebabkan karena algoritma TF-IDF menggunakan nilai desimal sementara CountVectorizer menggunakan nilai bulat. (sumber : [Mukesh Chaudhary, medium](https://medium.com/@cmukesh8688/tf-idf-vectorizer-scikit-learn-dbc0244a911a))